framework/include/cppmicroservices/detail/TrackedBundle.tpp

```diff
@@ -0,0 +1,136 @@
+/*=============================================================================
+
+  Library: CppMicroServices
+
+  Copyright (c) The CppMicroServices developers. See the COPYRIGHT
+  file at the top-level directory of this distribution and at
+  https://github.com/CppMicroServices/CppMicroServices/COPYRIGHT .
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+
+=============================================================================*/
+
+namespace cppmicroservices {
+
+namespace detail {
+
+template<class TTT>
+TrackedBundle<TTT>::TrackedBundle(BundleTracker<T>* bundleTracker,
+                                  BundleTrackerCustomizer<T>* customizer)
+  : Superclass(bundleTracker->d->context)
+  , bundleTracker(bundleTracker)
+  , customizer(customizer)
+  , latch{}
+{
+}
+
+template<class TTT>
+void TrackedBundle<TTT>::WaitOnCustomizersToFinish()
+{
+  latch.Wait();
+}
+
+template<class TTT>
+void TrackedBundle<TTT>::BundleChanged(const BundleEvent& event)
+{
+  // Call track or untrack based on state mask
+
+  // Ignore events that do not correspond with
+  // Bundle state changes
+
+  (void)latch.CountUp();
```

Comment:
This isn't the correct usage of the latch function. See DeclarativeServices CCActiveState.cpp for example of correct usage. 

 Example code:
 * function1()
 * {
 *    if(latch.CountUp())
 *    {
 *       // do something critical here
 *       latch.CountDown();
 *    }
 * }
 * function2()
 * {
 *    latch.Wait(); // this blocks the thread until all threads that have exited function1
 *    // do something here.
 * }
The CountUp function protects some code. That code can only be executed if CountUp returns true.

Reply:
My usage here is duplicated from a commit by Jeff to resolve a race in the `ServiceTracker` about 15 months ago where the tracker closes before the customizers are finished.

Right now, when `Close()` is called it calls `CounterLatch::Wait()`. As I understand it now, this should cause `CounterLatch::CountUp` to return false after that point, which we are not checking for currently and probably should be.

@jeffdiclemente any ideas on how to deal with this?

Reply:
in this instance `CountUp();` is used to ensure that all pending threads running `BundleChanged()` finish before the tracker cleans up and shuts down. There should be no need to check the return value of `CountUp()`. One could argue that a different synchronization mechanism should be used instead of the latch in this case.

Reply:
It is true that once the `Wait()` is called, we don't need to process any additional `BundleChanged()` events since the tracker is closing anyway. Definitely multiple ways to address this.

Reply:
No, once `Wait()` is called, it blocks the thread until the latch count reaches 0.
Closing the tracker unregisters the bundle listener, so no new bundle events can come into the tracker. the tracker then waits for all existing bundle listeners to complete.